cdap-watchdog/src/main/java/co/cask/cdap/logging/logbuffer/LogBufferReader.java

```diff
@@ -0,0 +1,171 @@
+/*
+ * Copyright © 2019 Cask Data, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package co.cask.cdap.logging.logbuffer;
+
+import co.cask.cdap.api.common.Bytes;
+import co.cask.cdap.logging.serialize.LoggingEventSerializer;
+import com.google.common.io.Closeables;
+
+import java.io.Closeable;
+import java.io.DataInputStream;
+import java.io.EOFException;
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.List;
+
+/**
+ * Reader to read log buffer files.
+ */
+public class LogBufferReader implements Closeable {
+  private static final String FILE_SUFFIX = ".buf";
+  private final int batchSize;
+  private final String baseDir;
+  private final long maxFileId;
+
+  private long currFileId;
+  private LogBufferInputStream inputStream;
+  private boolean skipFirstEvent;
+
+  public LogBufferReader(String baseDir, int batchSize, long currFileId, long currPos) throws IOException {
+    this.baseDir = baseDir;
+    this.batchSize = batchSize;
+    // if no checkpoints are written, start from 0th position from first file. Also in that case do not skip first event
+    this.currFileId = currFileId < 0 ? 0 : currFileId;
+    this.skipFirstEvent = currFileId >= 0;
```

Comment:
I think it would be better to base `skipFirstEvent` on `currPos` rather than `currFileId`. Because there may be a case where the `currFileId` gets incremented but the read loop exists before reading any line from the file.

Reply:
In the beginning when none of the log events are persisted, the offset value will be {fileId:-1, pos:-1}. In that case we want to recover logs from the beginning. However, when fileId and pos are non-negative, that indicates atleast one event is persisted. In that case first event should be skipped. This is little bit confusing. Added more comment to clarify.